### PR TITLE
Attempt flattening consequent OR and AND predicates in SqlPredicate

### LINKED ISSUE #7583: Long SqlPredicate gives StackOverflowException
If you have a very long SqlQuery with a lot of terms you will get StackOverflowException

    @Test
    public void testLongPredicate() {
        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance();
        IMap<String, String> map = hazelcastInstance.getMap(randomString());

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 4000; i++) {
            sb.append("this == " + i);
            sb.append(" or ");
        }

        sb.append(" this == -1");

        map.entrySet(new SqlPredicate(sb.toString()));
    }

The reason seems to be that the query is parsed as many nested OrPredicates rather than one OrPredicate with many terms.

### PR DESCRIPTION
**Pull request by vbekiaris**
While processing tokens in `SqlPredicate`, attempt to flatten consequent compound predicates (OR and AND). This is normally achieved with the `FlatteningVisitor` used by the rule-based query optimizer, however in long SQL predicates, current implementation nesting may result in `StackOverflowException` before we make it to the optimizer.
Fixes #7583

### CODE REVIEW

**Comment by jerrinot** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/util/collection/ArrayUtils.java [hunk 1]
```diff
@@ -82,4 +82,17 @@ private ArrayUtils() {
...
+
+    /**
+     * Copies in order {@code sourceFirst} and {@code sourceSecond} into {@code dest},
+     * i
+     * @param sourceFirst
+     * @param sourceSecond
+     * @param dest
+     * @param <T>
+     */
+    public static <T> void concat(T[] sourceFirst, T[] sourceSecond, T[] dest) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/util/collection/ArrayUtils.java]: `concat` can be tested in `ArrayUtilsTest`

File: hazelcast/src/test/java/com/hazelcast/query/SqlPredicateTest.java [hunk 2]
```diff
@@ -364,6 +374,114 @@ public void testInvalidSqlPredicate2() {
         new SqlPredicate("");
     }
 
+    // This test used to fail with a stack overflow exception, due to the way SQL predicates were
+    // nested, missing an opportunity to flatten.
+    // https://github.com/hazelcast/hazelcast/issues/7583
+    @Test
+    public void testLongPredicate() {
+        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance();
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/query/SqlPredicateTest.java]: `TestHazelcastInstanceFactory` can be used to create `hazelcastInstance `

**Comment by ahmetmircik** [comment]
👍 with minor comments

**Comment by vbekiaris** [comment]
run-lab-run

**Pull request merged**